book/src/types/rust_types/application_ty.md

```diff
@@ -22,5 +22,47 @@ _)`) or "fixed-length array" `[_; _]`. Note that the precise set of
 these built-in types is defined by the `Interner` and is unknown to
 chalk-ir.
 
+## [`TypeName`] variants
+
+### Generator
+
+A `Generator` represents a Rust generator. There are three major components
+to a generator:
+
+* Upvars - similar to closure upvars, they reference values outside of the generator,
+  and are stored across al yield points.
+* Resume/yield/return types - the types produced/consumed by various generator methods.
+  These are not stored in the generator across yield points - they are only
+  used when the generator is running.
+* Generator witness - see the `Generator Witness` section below.
+
+Of these types, only upvars and resume/yield/return are stored directly in
+`TypeName::Generator`. The generator witness is implicitly associated with the generator
+by virtue of sharing the same `GeneratorId`. It is only used when determining auto trait
+impls, where it is considered a 'constituent type'.
+
+### Generator witness types
+
+The `GeneratorWitness` variant represents the generator witness of
+the generator with id `GeneratorId`. 
```

Comment:
I think it'd be good to explain the *purpose* of a witness:

The witness type captures the types of all variables that may be live across a `yield` point.

Reply:
I just moved this entire block from a different section of the book :) - I'm happy to expand on it.